### PR TITLE
Warn about unknown command line arguments when starting NVDA

### LINKED ISSUE #12795: NVDA should warn about  unknown command line parameters
First of the series of issues created as a result of discussion in #12076

### Is your feature request related to a problem? Please describe.
Currently NVDA does not warn when it was started with invalid command line parameters. For example if someone misspells `--disable-addons` and attempts to rule that the given issue is caused by add-ons or not there is no indication that testing they're doing does not prove anything.
 This problem  exists because all CLI arguments unknown to NVDA are placed in `globalVars.appArgsExtra` (the motivation for this is to allow plugins to support additional CLI arguments)

### Describe the solution you'd like
NVDA should warn on startup about arguments which are unknown and which has not been used by add-ons. To do so it would be necessary to make plugins register for parameters they're interested in rather than just passively looking for them in `globalVars.appArgsExtra`.
To ensure that the new solution is as flexible as old `globalVars.appArgsExtra` were the following points need to be considered:
- First plugin which is interested in the given parameter does not remove it from the list of unknown params - the old system allowed two plugins to share parameter with the same name

The design which I have in mint is as follows:
-In addonHandler there is new extensionPoint `CLIParamIsKnown` to which all globalPlugins interested in processing command line parameters should register.
- Unfortunately none of the extensionsPoints we have currently would  work for this. `Filter` modifies data thereby removing a CLI param making it impossible for later plugin to process the same parameter) and `Decider` returns as soon as first interested plugin would be interested in the given parameter. The ideal option would be to create a `Decider` which accumulates results and returns `True` if at least one handler returns `True`, `False` otherwise.

After NVDA starts each argument placed in `globalVars.appArgsExtra` is provided to all handlers registered to the extension point - if at least one parameter is not known to add-ons `core.doStartupDialogs` should show a dialog informing about what unknown parameters were provided. The disadvantage of this approach is that all add-ons which want to parse extra command line parameters have to ship a global plugin (there is no guarantee that app module or other type of plugin would be loaded on startup).
### Describe alternatives you've considered
- Leave code as is
- Do not allow plugins to support command line parameters (I'm not aware about any which needs this functionality). @leonardder  Since `globalVars.appArgsExtra` were introduced by you in #6865 would you be able to tell if you know about any use cases or has this been implemented just in case some plugin may need this? I'll also ask about possible usages on add-ons list.

### PR DESCRIPTION
**Pull request by lukaszgo1**
### Link to issue number:
Closes #12795 
### Summary of the issue:
When starting NVDA with unknown (not used by NVDA nor by add-ons) 
 command line parameters there is no warning.
### Description of how this pull request fixes the issue:
1. Introduced a new extension point which accumulates decisions made by all handlers and allowed to register add-ons for command line arguments they want to use. Note that if the given add-on contains only app modules it has to bundle a small global plugin to make sure that it is loaded on NVDA's startup.
2. On NVDA startup user is warned if there are any arguments which are unknown to NVDA and not used by add-ons.
### Testing strategy:
- Started NVDA without unknown arguments
- Started NVDA with some unknown arguments and no add-ons interested in CLI arguments - made sure that   appropriate warning was shown and that all arguments were flagged as unknown
- Started NVDA with various combinations of known and unknown parameters and 
[these plugins](https://github.com/nvaccess/nvda/files/7589823/globalPlugins.zip) in the scratchpad - inspected log to made sure that each of the plugin had a chance to process all unknown parameters, and that only parameters not used by any of them were shown as unknown.
 
### Known issues with pull request:
None known
### Change log entries:
New features
- NVDA now warns about command line parameters which are unknown and not used by add-ons.
This can be potentially viewed as a change rather than new feature - I'm happy either way.
For Developers
- `globalVars.appArgsExtra` has been removed - if your add-on need to process additional command line arguments see documentation of `addonHandler.isCLIParamKnown` for details.
### Code Review Checklist:

- [X] Pull Request description:
  - description is up to date
  - change log entries
- [X] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [X] API is compatible with existing add-ons.
- [X] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [X] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

### CODE REVIEW

**Comment by lukaszgo1** [comment]
@feerrenrut  Given our discussion in #12795 you may be interested in reviewing this PR. It would be nice to add it to 2022.1 milestone as it breaks backwards compatibility. I decided against your proposal about forcing add-ons developers to use a known prefix for all arguments since while usages of additional parameters is not very common changing them would annoy users unnecessarily.

**Comment by lukaszgo1** [comment]
@josephsl  Given your SPL add-on is using `globalVars.appArgsExtra` currently I've taken a quick look at its code and while the solution proposed here would certainly allow you to parse additional command line parameters you need, there is some magic happening with `globalVars.appArgsExtra` in your code which is either wildly incorrect or I am misunderstanding your intent. Essentially you're discarding parameters  your add-on is interested in from `globalVars.appArgsExtra` at the termination of the app module for Studio and according to the comments this is done to avoid them being used on NVDA restart. This approach however does not work as intended since:
- Removing anything from `globalVars.appArgsExtra` does not mean these parameters are gone from `sys.argv` therefore they would still be used if NVDA is restarted
- Removing these arguments at the appModule termination results in a pretty confusing behavior for users - if they specified a CLI option and closed Studio the behavior is different after Studio is re-launched.

**Comment by josephsl** [comment]
Hi, I see. At the moment I have no plans to work on SPL add-on further as I’m ending maintenance for it. The intention was to remove add-on specific command-line arguments when the app module has terminated, but looks like I may need to look at it, or perhaps remove SPL specific command-line options altogether at the cost of removing testing facilities for some features. Thanks.

**Comment by josephsl** [comment]
Hi,

The change in behavior for Studio add-on between Studio sessions is intentional - I have done it so the command-line parameters affect NVDA during a single Studio session. The easiest workaround that would not involve massive reengineering would be telling NVDA to use add-on specific command-line switches as long as current NVDA session is active.

Also, keep in mind that Studio add-on is really a collection of app modules, so extra command-line switches will be active as long as the main Studio app module is active. If I understand the PR correctly, add-ons must tell NVDA that it needs specific parameters at startup, which may or may not be feasible for app module add-ons (certainly it will work for global plugins and drivers). One implication is that NVDA may treat a switch needed by an app module add-on as invalid until the app module loads, in which case NVDA will be told that the app module needs an invalid command-line switch or two (such is the case with Studio app module).

One workaround would be for app module add-ons to ship with a global plugin whose main task would be informing NVDA of command-line switches the app module requires, unless I'm misunderstanding the PR source code text correctly.

As for studio add-on, as soon as the PR is approved, I'll modify my add-on accordingly to handle old and new code paths.

Thanks.

**Comment by lukaszgo1** [comment]
You're right that for add-ons consisting of app modules it is necessary to bundle a global plugin which would be responsible for  parsing command line arguments - I've edited the PR description to  make it clear. If, for the SPL add-on the intention was indeed to use these arguments only for the first run of Studio, then source code comments should be edited and no longer talk about removing them to preserve default behavior after restart of NVDA as that is not what happens.

**Comment by josephsl** [comment]
Hi, that is what will happen with Studio add-on source code – I’m also making sure the extra argument is gone from sys.argv just to play safe. Thanks.

File: source/addonHandler/__init__.py
Comment (seanbudd) [inline source/addonHandler/__init__.py:49]:
```suggestion
# Each handler is called with one keyword argument `cliArgument`
```

File: source/addonHandler/__init__.py
Comment (seanbudd) [inline source/addonHandler/__init__.py:51]: Could you add an example of how an addon may use this extension point to the documentation here. 

For example, something like this:
```py
def processArgs(cliArgument: str) -> bool:
     if cliArgument == "enable-example-addon-feature":
            doSomething()
            return True
     return False

isCLIParamKnown.register(processArgs)
```

**Comment by seanbudd** [comment]
It would be helpful to add some system tests for this, using the example plugins, but I fear we may need to change the system infrastructure significantly to support this.

**Comment by lukaszgo1** [comment]
I believe all requested changes are now addressed though I've placed an example code in the developer guide rather than in `addonHandler` as IMO it is a better place for a code snippets like this.

> It would be helpful to add some system tests for this, using the example plugins, but I fear we may need to change the system test infrastructure significantly to support this.


I don't know enough about our system testing infrastructure to be able to asses how much work it would require, however for me the main blocker for working on system tests is the fact that they rely on Windows being in English therefore I cannot execute them on my main development machine. This is something which I intent to improve at some point but honestly it's low priority for me. Is this PR going to be accepted without system tests, or do you consider them required for getting this merged?

File: devDocs/developerGuide.t2t
Comment (seanbudd) [inline devDocs/developerGuide.t2t:704]:
```suggestion
Note that since command line arguments are processed just after NVDA starts, your add-on needs to process them in the global plugin, since app modules or other drivers may not be loaded at this stage.
```

File: devDocs/developerGuide.t2t
Comment (seanbudd) [inline devDocs/developerGuide.t2t:703]:
```suggestion
However if you want to use any additional arguments, this is possible by adding a handler to the extension point ``addonHandler.isCLIParamKnown``.
```

**Review by seanbudd — COMMENTED** [review]
I've suggested some minor rephrasing to the developerGuide, but this otherwise looks good to me.

>Is this PR going to be accepted without system tests, or do you consider them required for getting this merged?

The inclusion of system tests will not block this PR. 
I wanted to flag that this is another instance where system tests for command line arguments would be helpful. I am going to create an issue for adding them, as it has been discussed before and it is getting more important. Once the system test infrastructure supports testing command line arguments, they will be expected for PRs like this.

I am going to perform some extensive local testing before final approval.

**Review by seanbudd — APPROVED** [review]

**Pull request merged**
